ksqldb-engine/src/main/java/io/confluent/ksql/execution/ExecutionPlanBuilder.java

```diff
@@ -87,4 +97,142 @@ public ExecutionPlan buildPhysicalPlan(
         resultStream.getSourceStep()
     );
   }
+
+
+  private void optimizeForSelfJoin(final PlanNode currentNode) {
+
+    if (currentNode instanceof JoinNode) {
+      final JoinNode joinNode = (JoinNode) currentNode;
+      if (! matchSubGraphPattern1(joinNode)) {
```

Comment:
The optimization looks for the pattern: 

```
      ---> PreJoinProject ---> PreJoinRepartition ---> DataSource
Join
      ---> PreJoinProject ---> PreJoinRepartition ---> DataSource
```

If the DataSource is the same, then it is a self-join. However, from testing I found that the `PreJoinRepartition` nodes are not always present. Hence, we need to look for other patterns as well. Namely:

```
      ---> PreJoinProject  ---> DataSource
Join
      ---> PreJoinProject  ---> DataSource
```

and

```
      ---> DataSource
Join
       ---> DataSource
```
The latter should not occur but I am covering my bases. 

In order to make the code clearer and easier to understand and maintain, I broke each pattern into a different method that tries to find it.